### PR TITLE
[Connect ] Account as foregin key

### PR DESCRIPTION
**Pull request by ticosax**
#### What's this PR do?
Replace stripe_account from a `CharField` to a `ForeignKey`, because it helps to prevent data inconsistencies.

#### Definition of Done (check if considered and/or addressed):

- [x] Are all backwards incompatible changes documented in this PR?
- [x] Have all new dependencies been documented in this PR?
- [x] Has the appropriate documentation been updated (if applicable)?
- [x] Have you written tests to prove this change works (if applicable)?

- [x] update the migration to drop columns and add new ones.

### CODE REVIEW

File: pinax/stripe/webhooks.py [hunk 1]
```diff
@@ -97,7 +99,6 @@ def validate(self):
             )
         )
         self.event.valid = self.event.webhook_message["data"] == self.event.validated_message["data"]
-        self.event.save()
```
Comment (blueyed) [inline pinax/stripe/webhooks.py]: Why is this removed?

Comment (ticosax) [reply pinax/stripe/webhooks.py]: it is called again few lines below

Comment (ticosax) [reply pinax/stripe/webhooks.py]: in self.process()

**Comment by ticosax** [comment]
IMO this should be in Samwise, otherwise we won't be able to apply it as it will corrupt data.

**Comment by paltman** [comment]
I tend to agree but I wonder if we are missing something. @lukeburden I think this was your original model.  Any reason it's a `CharField` other than a `ForeignKey`?  Anything to be concerned about by moving to a FK?

**Comment by jtauber** [comment]
Is it for robustness in the case of account deletion?

I've certainly seen CharField instead of ForeignKey used for that reason elsewhere.

**Comment by lukeburden** [comment]
I left it as a CharField because I made it a CharField initially, and never found a reason to upgrade it to a ForeignKey.

And I'm not sure I see the reason to do this now, either:

- I'm not worried about data-inconsistencies, as this is handled by Stripe - this project is a cache, so my sense is that it's job isn't really to enforce integrity for foreign keys. Also, as long as we're syncing objects from Stripe-provided representations there is no real likelihood of getting stripe ids somehow and injecting bad data into the stripe_account field.
- If, for whatever reason (webhooks not being processed, Account deleted, etc) an Account does not exist in our database, with the use of a ForeignKey we actually fail to cache the association of an account related object to an account, as instead of just blindly storing the stripe_id of the account, we fail to find an Account ID and store null. Note - I run into race conditions quite frequently along these lines when receiving webhooks for Charges when somehow the Customer doesn't yet exist in the database.

Whoosh, that got lengthy - sorry. I think if we can guarantee the Account exists already somehow, it could be OK to go for a FK, but overall this change strikes me as adding complications without real  practical gain. Interested to hear your thoughts.

**Comment by ticosax** [comment]
The practical gain I see are
- better admin integration i.e:
  https://github.com/lock8/pinax-stripe/commit/81e47772371228429182932e9707e634e9f8bd9f
- performance:
because filtering by account could become slow in presence of large set of entries per account.
We would need to add some Index on the stripe_account column on some models, like `Customer`, to filter them by account. In the other hand by performing a join on the primary key, we already have the index when we need to filter resources by account. Decisions regarding adding this index or not will not be easy to make as not everyone will be satisfied; depending on each use case.

To solve the issue with an absent account that could prevent the event to be processed, since the only required field is stripe_id, `Account.objects.get_or_create(stripe_id=stripe_account_id)` Could be used in every case by the WebHook processor.

**Comment by blueyed** [comment]
This needs some more fixes, so please merge other PRs first.

**Comment by paltman** [comment]
@ticosax can you merge the migrations for this?

**Comment by paltman** [comment]
Thanks @blueyed.  I'm guessing this is ready now once the builds all pass.

**Comment by blueyed** [comment]
Yes, go for it.
